eth/vm/tracing.py

```diff
@@ -0,0 +1,246 @@
+        stack = computation.dump_stack() if self.is_storage_enabled else None
+        memory = computation.dump_memory() if self.is_memory_enabled else None
+
+        if self.is_storage_enabled:
+            storage_address = computation.msg.storage_address
+            if opcode_fn.mnemonic == "SSTORE" and len(stack) >= 2:
+                val = stack[len(stack) - 2]
+                slot = stack[len(stack) - 1]
+                self.changed_values.set_slot(storage_address, slot, val)
+            storage = self.changed_values.dump(storage_address)
+        else:
+            storage = None
+
+        try:
+            yield
+        except VMError as err:
+            self._log_operation(
+                depth=computation.msg.depth + 1,
+                error=err,
+                gas=start_gas,
+                gas_cost=start_gas - computation.get_gas_remaining(),
+                memory=memory,
+                op=opcode_fn.mnemonic,
+                pc=pc,
+                stack=stack,
+                storage=storage
+            )
+            raise
+        else:
+            self._log_operation(
+                depth=computation.msg.depth + 1,
+                error=None,
+                gas=start_gas,
+                gas_cost=start_gas - computation.get_gas_remaining(),
+                memory=memory,
+                op=opcode_fn.mnemonic,
+                pc=pc,
+                stack=stack,
+                storage=storage
+            )
+
+    def finalize(self, computation: 'BaseComputation') -> None:
+        if self.is_final:
+            raise ValidationError("Cannot finalize tracer which is already finalized")
+        elif computation.is_origin_computation:
+            self.result = ExecutionResult(
+                error=computation.error is not None,
+                gas=computation.get_gas_used(),
+                logs=tuple(self.logs),
+                output=computation.output,
+            )
+
+    def _log_operation(self,
+                       *,
+                       depth: int,
+                       error: Optional[VMError],
+                       gas: int,
+                       gas_cost: int,
+                       memory: bytes,
+                       op: str,
+                       pc: int,
+                       stack: Tuple[int, ...],
+                       storage: Dict[int, Union[int, bytes]]) -> None:
+        if self.is_full:
+            self.logger.debug(
+                'StructTracer full (limit=%d). Discarding trace log entry',
+                self.limit,
+            )
+            return
+
+        self.logs.append(StructLogEntry(
+            depth=depth,
+            err=error,
+            gas=gas,
+            gas_cost=gas_cost,
+            memory=memory,
+            op=op,
+            pc=pc,
+            stack=stack,
+            storage=storage,
+        ))
+
+
+def trace_transaction(chain: 'BaseAsyncChain', tx_hash: Hash32, memory: bool, stack: bool,
```

Comment:
This utility should just be moved into the `trinity` module somewhere near the `debug_traceTransaction` implementation.